### PR TITLE
[chore][CI/CD] Fix bugs in get-codeowners.sh

### LINKED ISSUE #29744: [CI/CD] Stale issues workflow pinging invalid components
### Component(s)

_No response_

### Describe the issue you're reporting

**Description**

The `Mark issue as stale` workflow has a bug in its logic when pinging code owners based on labels.

Here's the problem code:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95cb7336823185453e7a6538dc38988ee6596042/.github/workflows/scripts/mark-issues-as-stale.sh#L45-L48
The way this is written is labels are broken up into words, not kept together as a whole.

Example:
```
crobert$ ~/dev/opentelemetry-collector-contrib $ LABELS=$(gh issue view 27577 --json labels --jq '.labels.[].name')
crobert$ ~/dev/opentelemetry-collector-contrib $ for LABEL in ${LABELS}; do
> echo $LABEL
> done
bug
help
wanted
good
first
issue
Stale
flaky
test
receiver/sqlquery
receiver/oracledb
```

The script in question will then check for valid code owners for each "label" treating these split labels as possible components. The label `flaky test` is split into `flaky` and `test`, and then the get-codeowners.sh script is called on each word, `flaky`, and `test`. Before https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29572 the `get-codeowners` script would search for `test`, find multiple matches, then search `test/test`. This search would fail as there are no components named `test/test`. However, after https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29572, `get-codeowners` will search `test/` and return the first match as if it's a valid component. Currently, the first match is `internal/k8stest`, which will ping me as I'm the only code owner there.

Examples of this bug:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24773#issuecomment-1849268024
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27577#issuecomment-1849268082

**Possible solutions**
1. Add check in `get-codeowners.sh` to only check for codeowners if the given `COMPONENT` is a valid component as defined by the [get-components.sh](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95cb7336823185453e7a6538dc38988ee6596042/.github/workflows/scripts/get-components.sh) script. We should return empty code owners here and not exit unsuccessfully, as this would match current behavior of the the script. Also, I think this is a good idea as `get-components.sh` is what's used to [generate labels for this repository](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95cb7336823185453e7a6538dc38988ee6596042/.github/workflows/scripts/generate-component-labels.sh#L27C12-L27C12).
2. `get-codeowners.sh` could match from the beginning of lines, so the given `COMPONENT` has to match a line's start. This is valid as well because the `get-components.sh` script is matching the first column of the component rows, shown [here.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95cb7336823185453e7a6538dc38988ee6596042/.github/workflows/scripts/get-components.sh#L10)
3. `get-codeowners.sh` could match entire words (`-x` option in grep), not just text within a line.
4. Modify `Mark issue as stale` script to only ping on complete labels, not split labels.

I think a combination of 1, 2, and 4 would be best here. I don't see a time where a label should have multiple words corresponding to multiple components, so I think 4 makes sense. 1 and 2 are mostly validating functionality of the `get-codeowners.sh` script to make sure it properly handles invalid input. 3 may just be complicated or unclear.

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

There's a combination of bugs currently existing in `get-codeowners.sh` and the `Mark issues as stale` workflow. This implements proposed solution 2 of #29744
- Match component against beginning of line in code owners.

**Issues**
Resolves #29744

**Testing**
Basic spot checking script:
```
COMPONENTS=("azure" "processor/resourcedetection" "processor/resourcedetectionprocessor" "test" "testbed" "extension/observer" "internal/k8stest" "extension/encoding" "flaky" "pkg/translator" "tools" )

for EXAMPLE_COMPONENT in ${COMPONENTS[@]}; do
    OWNERS=$(COMPONENT=${EXAMPLE_COMPONENT} bash ".github/workflows/scripts/get-codeowners.sh")
    echo $EXAMPLE_COMPONENT: $OWNERS
done
```
Output:
```
azure:
processor/resourcedetection: @Aneurysm9 @dashpole
processor/resourcedetectionprocessor: @Aneurysm9 @dashpole
test:
testbed: @open-telemetry/collector-approvers
extension/observer: @dmitryax @rmfitzpatrick
internal/k8stest: @crobert-1
extension/encoding: @atoulme @dao-jun @dmitryax @MovieStoreGuy @VihasMakwana
flaky:
pkg/translator:
tools:
```
Reference [CODEOWNERS](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/.github/CODEOWNERS) as source of truth.

### CODE REVIEW

**Review by atoulme — APPROVED** [review]
LGTM ; pretty advanced bash-fu for this

File: .github/workflows/scripts/get-codeowners.sh
Comment (evan-bradley) [inline .github/workflows/scripts/get-codeowners.sh:28]:
```suggestion
VALID_COMPONENT_WITH_TYPE=$(bash "${CUR_DIRECTORY}/get-components.sh" | grep -x "${COMPONENT}${COMPONENT_TYPE}" || true)
```
Nit, but I'd like to stay consistent with using curly braces.

File: .github/workflows/scripts/get-codeowners.sh
Comment (evan-bradley) [inline .github/workflows/scripts/get-codeowners.sh:17]:
```suggestion
  echo "$((grep -m 1 "^${1}/ " .github/CODEOWNERS || true) | sed 's/   */ /g' | cut -f3- -d ' ')"
```
Could we solve this issue by making this regex more strict? I think if we add this (and remove the extra `/` on line 46) it might achieve the same effect.

**Review by evan-bradley — COMMENTED** [review]
Thank you for taking this on.

File: .github/workflows/scripts/get-codeowners.sh
Comment (crobert-1) [reply .github/workflows/scripts/get-codeowners.sh]: Updated as requested, I also simplified the if/else at the end of the script, let me know if it's more clear or if I should revert.

**Comment by crobert-1** [comment]
_Test results are the same after making changes requested_

File: .github/workflows/scripts/get-codeowners.sh
Comment (evan-bradley) [inline .github/workflows/scripts/get-codeowners.sh:17]:
```suggestion
  echo "$((grep -m 1 "^${1}/ " .github/CODEOWNERS || true) | \
```
Sorry, I should have explained my suggestion a little more in-depth. My thought is that if we use the regex `^${1}/ ` to match for the component, we can check that our search (`$1`) is bounded by the start of the line and the CODEOWNERS file delimiting character (a space), which I think should ensure we're not just matching on part of the name.

Do you think if we do this check that we could get away without doing additional validations on the component name below?

Comment (crobert-1) [reply .github/workflows/scripts/get-codeowners.sh:17]: My apologies, misread the suggestion as simply `/`, not `/ `. Do we have guarantees that indentantions in `CODEOWNERS` will always be a space and not some other whitespace character (like a tab? I'm not seeing anything definitive in GitHub documentation, but our file at least is all spaces after paths from what I can tell.

From testing we can get rid of the extra validation with this change. In my mind it's always nice to validate inputs upon entry, but if we're confident enough with the new grep statement I guess it's fair to remove.

Comment (evan-bradley) [reply .github/workflows/scripts/get-codeowners.sh]: That's a fair point, we could just do `^${1}/\s` to capture any other types of whitespace characters. The closest thing I can find to an authoritative guide on the syntax is [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-syntax), which uses spaces in their example, but the delimiting character isn't explicitly mentioned.

I'm mostly looking to cut down on code since shell scripting isn't the easiest to read, but if you think we may benefit from additional validation I'll defer to you here.

Comment (crobert-1) [reply .github/workflows/scripts/get-codeowners.sh]: I've added the whitespace regex, thanks!

We can remove the validation for now, then re-add again in the future if we find another case that we need to account for. 👍

**Review by evan-bradley — APPROVED** [review]
Thanks again for taking care of this.

**Pull request merged**
